charge/clients/huggingface_client.py

```diff
@@ -0,0 +1,266 @@
+            import torch
+        except ImportError:
+            raise ImportError(
+                "Please install transformers and torch: "
+                "pip install transformers torch"
+            )
+        
+        self._model_path = model_path
+        self._model_info = model_info or {
+            "vision": False,
+            "function_calling": True,
+            "json_output": True,
+            "family": ModelFamily.UNKNOWN,
+            "structured_output": True,
+        }
+        
+        # Convert string dtype to torch dtype
+        dtype_map = {
+            "auto": "auto",
+            "float16": torch.float16,
+            "bfloat16": torch.bfloat16,
+            "float32": torch.float32,
+        }
+        torch_dtype_obj = dtype_map.get(torch_dtype, "auto")
+        
+        quantization_config = None
+        
+        # Load tokenizer
+        self._tokenizer = AutoTokenizer.from_pretrained(
+            model_path,
+            trust_remote_code=trust_remote_code,
+            **kwargs
+        )
+        
+        # Prepare model loading arguments
+        model_kwargs = {
+            "device_map": device,
+            "trust_remote_code": trust_remote_code,
+            "torch_dtype": torch_dtype_obj,
+        }
+        
+        # Load model
+        self._model = AutoModelForCausalLM.from_pretrained(
+            model_path,
+            **model_kwargs,
+            **kwargs
+        )
+
+        # Set pad token if not set
+        if self._tokenizer.pad_token is None:
+            self._tokenizer.pad_token = self._tokenizer.eos_token
+    
+    async def create(
+        self,
+        messages: List[Any],
+        **kwargs
+    ) -> Any:
+        """
+        Create a completion from messages.
+        
+        Args:
+            messages: List of message dicts with 'role' and 'content'
+            **kwargs: Additional generation parameters
+        """
+        import asyncio
+        
+        # Run inference in thread pool to avoid blocking event loop
+        def _generate():
+            try:
+                # Convert messages to prompt
+                prompt = self._format_messages(messages)
+                
+                # Tokenize with proper max_length
+                max_length = getattr(self._model.config, 'max_position_embeddings', 2048)
+                inputs = self._tokenizer(
+                    prompt,
+                    return_tensors="pt",
+                    padding=True,
+                    truncation=True,
+                    max_length=max_length - 512,  # Leave room for generation
+                ).to(self._model.device)
+        
+                # Generate
+                gen_kwargs = {
+                    "max_new_tokens": kwargs.get("max_tokens", 4096),
+                    "temperature": kwargs.get("temperature", 0.7),
+                    "top_p": kwargs.get("top_p", 0.9),
+                    "do_sample": kwargs.get("temperature", 0.7) > 0,
```

Comment:
isn't it better to not give those arguments at all, if not provided, so as to let the model config json files decide what to set them to?

Reply:
I think that we should refactor this later to make these proper env variables.